sensors/Dht.py

```diff
@@ -0,0 +1,98 @@
+# inheritance library
+# import abc
+import Adafruit_DHT
+# GPIO lib
+import RPi.GPIO as GPIO           # import RPi.GPIO module
+
+from sensors.SensorException import SensorReadError
+
+
+# SensorInterface(Top Hierarchy)
+# Author     : clintonbf
+# Repository :   https://github.com/clintonbf/sensor_interface/blob/master/sensors/SensorInterface.py
+from sensors.Sensor import Sensor
+
+# unit conversion function
+# Author     : clintonbf
+# Repository : https://github.com/clintonbf/sensor_interface/tree/master/unit_conversion
+from utils.unit_conversion.temperature import celsius_to_kelvin
+
+DATA_INDICES = {
+    "relative_humidity": 0,
+        "temperature": 1,
+    }
+
+class Dht(Sensor):
+    """
+        Credit for connect_to_port and get_raw_data goes to
+        Noah MacRitchie (noah21mac@gmail.com) and Andrey Goryelov (andrey.goryelov@gmail.com)
+    """
+
+    # TODO: what is timeout ?
+    # TODO: It is just a GPIO data in, not serial connection.
+    # TODO: pass it in like : gpio_in = 1, named params
+   
+    def __init__(self, uid: int, gpio_in: str, model : str ) -> None:
+        # super().__init__()
+        self.__uid = uid
+        self.__gpio_in = gpio_in
+        # self.__timeout = timeout
+        if model.lower() == 'dht11':
+            self.__model = 11
+        # elif model.lower() == 'am2302':
+        #     self.__model = 22
+        else:
+            self.__model = 22
+        self.__reading = () # TODO meaning?
+
+    @classmethod
+    def __subclasshook__(cls, subclass):
+        return(hasattr(subclass, 'format_data') and
+               callable(subclass.format_data) or
+               NotImplemented)
+
+    # Sets the GPIO to be in input mode
+    def connect_to_sensor(self) -> int:
+        DHT_in = self.__gpio_in
+        # TODO return a value for error checking?
+        GPIO.setmode(GPIO.BCM)                 # choose BCM or BOARD
+        status = GPIO.setup(DHT_in, GPIO.IN)  # set a port/pin as an input
+        return status
+
+    # Take reading from the sensor
+    # (with retry, it's garunteed to get an output with up to 15 trials )
+    def take_reading(self) -> int:
+        try:
+            self.__reading = Adafruit_DHT.read_retry(self.__model, self.__gpio_in)
+        except:
+            raise SensorReadError("cannot get a reading after 15 trials, please check wiring")
+
+
+    # TODO: necessary?
+    def get_raw_data(self)-> float:
+        DHT_in = self.__gpio_in   
+        GPIO.setmode(GPIO.BCM)  # choose BCM or BOARD
+        temp, humidity = Adafruit_DHT.read_retry(self.__model, DHT_in)
+
+    def get_temperature(self) -> float:
+        temperature_reading = 0.00
+        try:
+             temperature_reading = celsius_to_kelvin(self.__reading[DATA_INDICES['temperature']])
+        except:
```

Comment:
This exception does not belong here. Remove it

Reply:
The Adafruit_DHT.read_retry() function will return (None,None) after 15 trials if nothing is read from the sensor,
so I think it makes sense to categorize it as a SensorReadError.

Reply:
I don't think it should go in a getter.

It would be more appropriate to put it in get_raw_data().

Reply:
I'll move the SensorReadError into get_raw_data and introduce a new error type called "Unit conversion error", which I feel is needed in this circumstance.

Reply:
So, if the data passed in is None then an exception is raised?

That makes more sense to me.

Reply:
Yes, a UnitConversionError is raised.